File: frontend/src/utilidades/validacion.js
const NOMBRE_VALIDO = /^[\p{L}]+(?:[ '-][\p{L}]+)*$/u;
const TELEFONO_VALIDO = /^\+?[0-9][0-9 ()-]*[0-9]$/;
const CORREO_VALIDO = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const CARACTER_ESPECIAL = /[^\p{L}\p{N}_\s]/u;
const CONTRASENAS_DEBILES = new Set([
  "12345678",
  "admin123*",
  "contraseña123*",
  "password123*",
  "psicoconecta123*",
  "qwerty123*",
]);

export const AYUDA_CONTRASENA =
  "Usa de 8 a 15 caracteres con mayúscula, minúscula, número y carácter especial.";

export const obtenerRequisitosContrasena = (valor = "") => [
  {
    id: "longitud",
    texto: "Entre 8 y 15 caracteres",
    cumplido: valor.length >= 8 && valor.length <= 15,
  },
  {
    id: "mayuscula",
    texto: "Al menos una mayúscula",
    cumplido: /[A-ZÁÉÍÓÚÜÑ]/.test(valor),
  },
  {
    id: "minuscula",
    texto: "Al menos una minúscula",
    cumplido: /[a-záéíóúüñ]/.test(valor),
  },
  {
    id: "numero",
    texto: "Al menos un número",
    cumplido: /\d/.test(valor),
  },
  {
    id: "especial",
    texto: "Al menos un carácter especial",
    cumplido: CARACTER_ESPECIAL.test(valor),
  },
  {
    id: "sin-espacios",
    texto: "Sin espacios",
    cumplido: valor.length > 0 && !/\s/.test(valor),
  },
  {
    id: "no-comun",
    texto: "No debe ser una contraseña común",
    cumplido:
      valor.length > 0 &&
      !CONTRASENAS_DEBILES.has(valor.toLowerCase()) &&
      new Set(valor.toLowerCase()).size >= 5,
  },
];

export const normalizarEspacios = (valor = "") =>
  valor.trim().replace(/\s+/g, " ");

export const normalizarCorreo = (valor = "") => valor.trim().toLowerCase();

export const validarNombre = (valor, etiqueta) => {
  const normalizado = normalizarEspacios(valor);
  if (!normalizado) return `${etiqueta} es obligatorio.`;
  if (normalizado.length < 2) return `${etiqueta} debe tener al menos 2 caracteres.`;
  if (normalizado.length > 80) return `${etiqueta} no puede superar los 80 caracteres.`;
  if (!NOMBRE_VALIDO.test(normalizado)) {
    return "Usa únicamente letras, espacios, apóstrofes o guiones.";
  }
  return "";
};

export const validarCorreo = (valor) => {
  const normalizado = normalizarCorreo(valor);
  if (!normalizado) return "El correo electrónico es obligatorio.";
  if (normalizado.length > 255 || !CORREO_VALIDO.test(normalizado)) {
    return "Ingresa un correo electrónico válido.";
  }
  return "";
};

export const validarTelefono = (valor) => {
  const normalizado = normalizarEspacios(valor);
  if (!normalizado) return "";
  const digitos = normalizado.replace(/\D/g, "");
  if (
    normalizado.length > 20 ||
    digitos.length < 7 ||
    digitos.length > 15 ||
    !TELEFONO_VALIDO.test(normalizado)
  ) {
    return "Ingresa un teléfono válido de 7 a 15 dígitos.";
  }
  return "";
};

export const obtenerFechaMaximaAdulto = (edadMinima = 17) => {
  const fecha = new Date();
  fecha.setFullYear(fecha.getFullYear() - edadMinima);
  return fecha.toISOString().slice(0, 10);
};

export const validarFechaNacimiento = (valor, rol = "PATIENT") => {
  if (!valor) return "La fecha de nacimiento es obligatoria.";
  const fecha = new Date(`${valor}T00:00:00`);
  if (Number.isNaN(fecha.getTime())) {
    return "Ingresa una fecha de nacimiento valida.";
  }
  
  const edadMinima = rol === "PSYCHOLOGIST" ? 20 : 17;
  
  const fechaMinima = new Date();
  fechaMinima.setFullYear(fechaMinima.getFullYear() - 120);
  fechaMinima.setHours(0, 0, 0, 0);
  
  const fechaMaxima = new Date();
  fechaMaxima.setFullYear(fechaMaxima.getFullYear() - edadMinima);
  fechaMaxima.setHours(0, 0, 0, 0);

  if (fecha > fechaMaxima) return `Debes tener al menos ${edadMinima} años.`;
  if (fecha < fechaMinima) return "Ingresa una fecha de nacimiento valida.";
  return "";
};

export const validarContrasena = (valor) => {
  if (!valor) return "La contraseña es obligatoria.";
  if (
    valor.length < 8 ||
    valor.length > 15 ||
    /\s/.test(valor) ||
    !/[A-ZÁÉÍÓÚÜÑ]/.test(valor) ||
    !/[a-záéíóúüñ]/.test(valor) ||
    !/\d/.test(valor) ||
    !CARACTER_ESPECIAL.test(valor) ||
    CONTRASENAS_DEBILES.has(valor.toLowerCase()) ||
    new Set(valor.toLowerCase()).size < 5
  ) {
    return AYUDA_CONTRASENA;
  }
  return "";
};

export const validarPasswordRequerido = (valor) => {
  if (!valor) return "La contraseña es obligatoria.";
  if (valor.length > 15) return "La contraseña no puede superar los 15 caracteres.";
  return "";
};

export const obtenerErroresApi = (excepcion) => {
  const errores = excepcion.response?.data?.errors || {};
  return Object.fromEntries(
    Object.entries(errores).map(([campo, mensajes]) => [
      campo,
      Array.isArray(mensajes) ? mensajes[0] : mensajes,
    ]),
  );
};

export const obtenerMensajeApi = (excepcion, mensajeAlternativo) =>
  excepcion.response?.data?.message || mensajeAlternativo;

File: backend/servicios/servicio-usuarios/aplicacion/utilidades/validacion.py
import re

from marshmallow import ValidationError


NAME_PATTERN = re.compile(r"^[^\W\d_]+(?:[ '-][^\W\d_]+)*$", re.UNICODE)
PHONE_PATTERN = re.compile(r"^\+?[0-9][0-9 ()-]*[0-9]$")
ROLE_PATTERN = re.compile(r"^[A-Za-z][A-Za-z0-9_-]*$")
SPECIAL_CHARACTER_PATTERN = re.compile(r"[^\w\s]", re.UNICODE)
WEAK_PASSWORDS = {
    "12345678",
    "admin123*",
    "contraseña123*",
    "password123*",
    "psicoconecta123*",
    "qwerty123*",
}
PASSWORD_MESSAGE = (
    "La contraseña debe tener entre 8 y 128 caracteres, una mayúscula, "
    "una minúscula, un número y un carácter especial, sin espacios."
)


def normalize_spaces(value):
    return " ".join(value.strip().split()) if isinstance(value, str) else value


def normalize_email(value):
    return value.strip().lower() if isinstance(value, str) else value


def normalize_phone(value):
    if value is None:
        return None
    value = normalize_spaces(value)
    return value or None


def validate_name(value):
    if not NAME_PATTERN.fullmatch(value):
        raise ValidationError(
            "Usa únicamente letras, espacios, apóstrofes o guiones."
        )


def validate_phone(value):
    if value is None:
        return
    digits = re.sub(r"\D", "", value)
    if (
        not PHONE_PATTERN.fullmatch(value)
        or not 7 <= len(digits) <= 15
        or len(value) > 20
    ):
        raise ValidationError(
            "Ingresa un teléfono válido de 7 a 15 dígitos."
        )


def validate_password(value):
    if (
        not 8 <= len(value) <= 128
        or any(character.isspace() for character in value)
        or not any(character.isupper() for character in value)
        or not any(character.islower() for character in value)
        or not any(character.isdigit() for character in value)
        or not SPECIAL_CHARACTER_PATTERN.search(value)
        or value.lower() in WEAK_PASSWORDS
        or len(set(value.lower())) < 5
    ):
        raise ValidationError(PASSWORD_MESSAGE)


def validate_role_name(value):
    if not ROLE_PATTERN.fullmatch(value):
        raise ValidationError(
            "Usa únicamente letras, números, guiones o guiones bajos."
        )
